opal/mca/pmix/pmix2x/pmix2x_client.c

```diff
@@ -598,12 +567,112 @@ int pmix2x_getnb(const opal_process_name_t *proc, const char *key,
     }
 
     /* call the library function */
-    rc = PMIx_Get_nb(&op->p, key, op->info, op->sz, val_cbfunc, op);
+    rc = PMIx_Get_nb(&op->p, pt->msg, op->info, op->sz, val_cbfunc, op);
     if (PMIX_SUCCESS != rc) {
         OBJ_RELEASE(op);
+        if (NULL != pt->valcbfunc) {
+            pt->valcbfunc(pmix2x_convert_rc(rc), NULL, pt->cbdata);
+        }
+        if (pt->lock.active) {
+            pt->status = pmix2x_convert_rc(rc);
+            OPAL_PMIX_WAKEUP_THREAD(&pt->lock);
+            return;
+        }
     }
 
-    return pmix2x_convert_rc(rc);
+    if (!pt->lock.active) {
+        OBJ_RELEASE(pt);
+    }
+}
+
+int pmix2x_get(const opal_process_name_t *proc, const char *key,
+               opal_list_t *info, opal_value_t **val)
+{
+    int rc;
+    pmix2x_threadshift_t pt;
+
+    opal_output_verbose(1, opal_pmix_base_framework.framework_output,
+                        "%s pmix2x:client get on proc %s key %s",
+                        OPAL_NAME_PRINT(OPAL_PROC_MY_NAME),
+                        (NULL == proc) ? "NULL" : OPAL_NAME_PRINT(*proc), key);
+
+    if (NULL == proc) {
+        /* if they are asking for our jobid, then return it */
+        if (0 == strcmp(key, OPAL_PMIX_JOBID)) {
+            (*val) = OBJ_NEW(opal_value_t);
+            (*val)->type = OPAL_UINT32;
+            (*val)->data.uint32 = OPAL_PROC_MY_NAME.jobid;
+            return OPAL_SUCCESS;
+        }
+        /* if they are asking for our rank, return it */
+        if (0 == strcmp(key, OPAL_PMIX_RANK)) {
+            (*val) = OBJ_NEW(opal_value_t);
+            (*val)->type = OPAL_INT;
+            (*val)->data.integer = pmix2x_convert_rank(my_proc.rank);
+            return OPAL_SUCCESS;
+        }
+    }
+
+    /* go ahead and threadshift anything else */
+    OBJ_CONSTRUCT(&pt, pmix2x_threadshift_t);
+    pt.source = proc;
+    pt.info = info;
+    pt.msg = key;
+    pt.valcbfunc = getcbfunc;
+    pt.cbdata = &pt;
+    OPAL_PMIX2X_THREADSHIFT(&pt, _get);
+    OPAL_PMIX_WAIT_THREAD(&pt.lock);
+    rc = pt.status;
+    *val = pt.val;
+    OBJ_DESTRUCT(&pt);
```

Comment:
I've seen a crash (just one in many, many runs) where at the very bottom of `_get` we call:
```c
if (!pt->lock.active) {
    OBJ_RELEASE(pt);
}
```

At the exact moment as `OBJ_DESTRUCT(&pt)` here and both destructors become active at the same time. Leaving to a crash in free as two threads fight over which one gets to free the memory. Should we move these OBJ_CONSTRUCT/OBJ_DESTRUCT to OBJ_NEW/OBJ_RETAIN/OBJ_RELEASE?